consensus/ethash/ethash.go

```diff
@@ -112,12 +112,13 @@ func memoryMapFile(file *os.File, write bool) (mmap.MMap, []uint32, error) {
 	if err != nil {
 		return nil, nil, err
 	}
-	// Yay, we managed to memory map the file, here be dragons
-	header := *(*reflect.SliceHeader)(unsafe.Pointer(&mem))
-	header.Len /= 4
-	header.Cap /= 4
-
-	return mem, *(*[]uint32)(unsafe.Pointer(&header)), nil
+	// The file is now memory-mapped. Create a []uint32 view of the file.
+	var view []uint32
+	header := (*reflect.SliceHeader)(unsafe.Pointer(&view))
+	header.Data = (*reflect.SliceHeader)(unsafe.Pointer(&mem)).Data
+	header.Cap = len(mem) / 4
+	header.Len = header.Cap
+	return mem, view, nil
```

Comment:
The other similar conversions seem a bit different. Perhaps lets do this part the same way? 

```
	var cache []byte
	cacheHdr := (*reflect.SliceHeader)(unsafe.Pointer(&cache))
	dstHdr := (*reflect.SliceHeader)(unsafe.Pointer(&dest))
	cacheHdr.Data = dstHdr.Data
	cacheHdr.Len = dstHdr.Len * 4
	cacheHdr.Cap = dstHdr.Cap * 4
```

Reply:
I think there is no real difference between what you posted and what's in the diff.
In the `[]uint32` to `[]byte` conversion in algorithm.go, it accesses the length of the
slice header instead of calling `len`. But they give the same result, and I prefer using `len`.

Reply:
I guess the difference is that the alternative uses `dstHdr` to retrieve the Len and Cap (no idea where these come from), vs you assume these will have certain values based on `mem`. I'm almost certain your code is fine since we never append to these constructed slices, but unsure what would hapen if we did.